### PR TITLE
Run3 GEMGeometryBuilder Run3 modifier 

### PR DESCRIPTION
**Pull request by hyunyong**
@jshlee @watson-ij 

GEM Run3 modifier setup for the alignment.

#### PR description:
Update the GEM geometry builder to use alignment tags.
Need to backport to 12_4_X

GEM doesn't have ideal geometry cff for digi.
Do we need to add ideal geometry cff and update configurations?
Some configurations use GEM default geometry builder as ideal geometry but they don't use the Run3 modifier, so it looks fine but I need to get a confirmation.

### CODE REVIEW

**Comment by tvami** [comment]
@hyunyong please make it clear in the PR description that you want this in 12_4_X...

**Comment by tvami** [comment]
type gem,new-feature

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by tvami** [comment]
This will fail w/o a new GT

File: Geometry/GEMGeometryBuilder/python/gemGeometryDB_cfi.py [hunk 1]
```diff
@@ -10,3 +10,7 @@
     alignmentsLabel = cms.string(''),
     applyAlignment = cms.bool(False)
 )
+
+from Configuration.Eras.Modifier_run3_GEM_cff import run3_GEM
+
+run3_GEM.toModify(GEMGeometryESModule, applyAlignment = True)
```
Comment (tvami) [inline Geometry/GEMGeometryBuilder/python/gemGeometryDB_cfi.py:16]: Done we also need
```
phase2_GEM.toModify(GEMGeometryESModule, applyAlignment = True)
```
as well?

Comment (francescobrivio) [reply Geometry/GEMGeometryBuilder/python/gemGeometryDB_cfi.py:16]: After private discussion offline between PR author and @cms-sw/alca-l2 it was decided to postpone the update of phase2 to a later moment.

**Comment by francescobrivio** [comment]
test parameters:
 - pull_request = #38692

**Comment by francescobrivio** [comment]
@cmsbuild please test

File: Geometry/GEMGeometryBuilder/python/gemGeometry_cff.py [hunk 2]
```diff
@@ -2,4 +2,8 @@
 
 from Configuration.ProcessModifiers.dd4hep_cff import dd4hep
 
+from Configuration.Eras.Modifier_run3_GEM_cff import run3_GEM
+
 dd4hep.toModify(gemGeometry, fromDDD = False, fromDD4hep = True)
+
+run3_GEM.toModify(gemGeometry, applyAlignment = True)
```
Comment (davidlange6) [inline Geometry/GEMGeometryBuilder/python/gemGeometry_cff.py:9]: naively this modification (and the dd4hep line above, but thats not new) should instead be in gemGeometry_cfi.py to keep the module definition in one place.

**Comment by francescobrivio** [comment]
This PR is changing the Run3 modifier only, but I see that phase2 GT and era are used for one of the failing jobs (command from [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-385255/26149/runTheMatrix-results/23234.0_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal/cmdLog)):
```
cmsDriver.py step2  \
  -s DIGI:pdigi_valid,L1TrackTrigger,L1,DIGI2RAW,HLT:@fake2 \
 --conditions auto:phase2_realistic_T15 \
 --datatier GEN-SIM-DIGI-RAW -n 10 --eventcontent FEVTDEBUGHLT \
 --geometry Extended2026D49 \
 --era Phase2C9 \
 --customise Validation/Performance/TimeMemoryJobReport.customiseWithTimeMemoryJobReport \
 --filein  file:step1.root  \
 --fileout file:step2.root  --suffix "-j JobReport2.xml " 
```
So i'm not sure I understand why this is failing...? Is Run3 geometry importend in phase2 eras?

**Comment by mmusich** [comment]
@francescobrivio 

> So i'm not sure I understand why this is failing...? Is Run3 geometry importend in phase2 eras?

Phase-2 inherits from Run-3

https://github.com/cms-sw/cmssw/blob/14a1fabf933e42874fd798b3b51d673131f71985/Configuration/Eras/python/Era_Phase2_cff.py#L21

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by francescobrivio** [comment]
@cmsbuild please abort

 - tests were started before we had change to update the GTs in the other PR

**Comment by francescobrivio** [comment]
@cmsbuild please test

 - now that the GTs in https://github.com/cms-sw/cmssw/pull/38692 have been updated the unitTests should not fail anymore

File: Geometry/GEMGeometryBuilder/python/gemGeometry_cff.py (hunk 2 above)
Comment (makortel) [reply Geometry/GEMGeometryBuilder/python/gemGeometry_cff.py:10]: An established pattern for such cases would be to have the ESModule to generate `gemGeometryDefault_cfi.py` and then to have a hand-written `gemGeometry_cfi.py` along
```py
from Geometry.GEMGeometryBuilder.gemGeometryDefault_cfi import gemGeometryDefault as _gemGeometryDefault

gemGeometry = _gemGeometryDefault.clone()

dd4hep.toModify(gemGeometry, fromDDD = False, fromDD4hep = True)
run3_GEM.toModify(gemGeometry, applyAlignment = True)
```
(such a reorganization may be outside of the scope of this PR)

**Comment by francescobrivio** [comment]
The failure in unit tests (despite the update of the GTs) is understood, i'm preparing another PR to fix it.
I will re-trigger the tests once the PR is ready.

**Comment by francescobrivio** [comment]
test parameters:
 - pull_requests = #38712,#38692

**Comment by francescobrivio** [comment]
@cmsbuild please test

 - third time's the charm

**Comment by francescobrivio** [comment]
Well the failure in the unitTest is now in the gem DQM client.
The stack trace is here: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-385255/26176/unitTests/src/DQM/Integration/test/TestDQMOnlineClient-gem_dqm_sourceclient/testing.log
but it's beyond my comprehension... 😞

**Comment by makortel** [comment]
This part
```
#3  0x00002ae4becd4a0b in sig_dostack_then_abort () from /cvmfs/cms-ib.cern.ch/nweek-02741/el8_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/el8_amd64_gcc10/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x00002ae4b70ab804 in edata_arena_ind_get (edata=<optimized out>) at include/jemalloc/internal/edata.h:258
...
#19 0x00002ae4b70b24c9 in fallback_impl<false> (size=512) at src/jemalloc_cpp.cpp:98
#20 0x00002ae517e5a847 in void std::vector<std::pair<int, std::pair<std::pair<float, float>, bool> >, std::allocator<std::pair<int, std::pair<std::pair<float, float>, bool> > > >::_M_realloc_insert<>(__gnu_cxx::__normal_iterator<std::pair<int, std::pair<std::pair<float, float>, bool> >*, std::vector<std::pair<int, std::pair<std::pair<float, float>, bool> >, std::allocator<std::pair<int, std::pair<std::pair<float, float>, bool> > > > >) () from /cvmfs/cms-ib.cern.ch/nweek-02741/el8_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/el8_amd64_gcc10/libDQMGEM.so
#21 0x00002ae517e554e5 in GEMDQMBase::readGeometryPhiInfoChamber(GEMStation const*, GEMDQMBase::MEStationInfo&) () from /cvmfs/cms-ib.cern.ch/nweek-02741/el8_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/el8_amd64_gcc10/libDQMGEM.so
```
hints towards this line
https://github.com/cms-sw/cmssw/blob/8303a585b01327b949afced21909dfc1443f874d/DQM/GEM/src/GEMDQMBase.cc#L342
in a case where the `emplace_back()` causes reallocation. Just guessing, could the loop
https://github.com/cms-sw/cmssw/blob/8303a585b01327b949afced21909dfc1443f874d/DQM/GEM/src/GEMDQMBase.cc#L314
have a "strange" number of iterations? Hard to say why it would cause a crash in jemalloc. An alternative could be that something before mismanages memory or corrupts jemalloc's data structures.

**Comment by quark2** [comment]
Hi, 

I'm one of the GEM onlineDQM developers, and I'm looking at this issue.

Best regards, 
Byeonghak Ko

**Comment by quark2** [comment]
Hi, 

I'm trying to reproduce the issue, but I got stuck by another issue. I met this error message.

```
----- Begin Fatal Exception 16-Jul-2022 14:24:53 CEST-----------------------
An exception of category 'NoRecord' occurred while
   [0] Processing global begin Run run: 344518
   [1] Prefetching for module GEMEffByGEMCSCSegmentSource/'gemEffByGEMCSCSegmentSource'
   [2] Calling method for EventSetup module GEMGeometryESModule/''
   [3] While getting dependent Record from Record MuonGeometryRecord
Exception Message:
No "GEMAlignmentRcd" record found in the EventSetup.

 Please add an ESSource or ESProducer that delivers such a record.
----- End Fatal Exception -------------------------------------------------
```

This is not the same as the error from the cmsbuild test [1].

My recipe is here: 

```
# Do not forget voms-proxy-init -voms cms -rfc

export SCRAM_ARCH=slc7_amd64_gcc10
cmsrel CMSSW_12_5_X_2022-07-08-2300
cd CMSSW_12_5_X_2022-07-08-2300/src
cmsenv
git cms-init -q
git cms-merge-topic hyunyong:GEMGeometryBuilder_CMSSW_12_5_X_2022-07-08-2300_PR
git cms-addpkg DQM/GEM
git cms-addpkg DQM/Integration
scram b -j 6

cmsRun DQM/Integration/python/clients/gem_dqm_sourceclient-live_cfg.py unitTest=True
```
(According to the name of the branch, I use `CMSSW_12_5_X_2022-07-08-2300` instead of `CMSSW_12_5_0_pre3`; the difference between this branch and `CMSSW_12_5_0_pre3` is already massive.)

Did I miss something? Please let me know anything I have to consider for the reproduce.

[1] https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-385255/26176/unitTests/src/DQM/Integration/test/TestDQMOnlineClient-gem_dqm_sourceclient/testing.log

**Comment by tvami** [comment]
@quark2 I think you need these two PRs as well:
https://github.com/cms-sw/cmssw/pull/38666#issuecomment-1182512524

**Comment by quark2** [comment]
Thanks, @tvami, (and sorry for my late...) I've added the branches and got a '`<signal handler called>`'. But it's different from the error messages on the log. This is what I got.

```
#0  0x00007faa12744ddd in poll () from /lib64/libc.so.6
#1  0x00007faa0341573f in full_read.constprop () from /cvmfs/cms-ib.cern.ch/nweek-02741/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/slc7_amd64_gcc10/pluginFWCoreServicesPlugins.so
#2  0x00007faa034160cc in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms-ib.cern.ch/nweek-02741/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/slc7_amd64_gcc10/pluginFWCoreServicesPlugins.so
#3  0x00007faa03418a1b in sig_dostack_then_abort () from /cvmfs/cms-ib.cern.ch/nweek-02741/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/slc7_amd64_gcc10/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x00007faa12687387 in raise () from /lib64/libc.so.6
#6  0x00007faa12688a78 in abort () from /lib64/libc.so.6
#7  0x00007faa126801a6 in __assert_fail_base () from /lib64/libc.so.6
#8  0x00007faa12680252 in __assert_fail () from /lib64/libc.so.6
#9  0x00007faa03f67a19 in dqm::implementation::DQMStore::initLumi(unsigned int, unsigned int, unsigned long) () from /cvmfs/cms-ib.cern.ch/nweek-02741/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/slc7_amd64_gcc10/libDQMServicesCore.so
#10 0x00007fa9a7610dc2 in DQMEDAnalyzer::beginRun(edm::Run const&, edm::EventSetup const&) () from /afs/cern.ch/work/b/bko/Work/dqm/test/CMSSW_12_5_X_2022-07-12-1100/lib/slc7_amd64_gcc10/pluginDQMGEMPlugins.so
#11 0x00007faa151ee035 in edm::stream::ProducingModuleAdaptorBase<edm::stream::EDProducerBase>::doStreamBeginRun(edm::StreamID, edm::RunTransitionInfo const&, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02741/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/slc7_amd64_gcc10/libFWCoreFramework.so
#12 0x00007faa151cb6f0 in edm::WorkerT<edm::stream::EDProducerAdaptorBase>::implDoStreamBegin(edm::StreamID, edm::RunTransitionInfo const&, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02741/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/slc7_amd64_gcc10/libFWCoreFramework.so
#13 0x00007faa150c4709 in decltype ({parm#1}()) edm::convertException::wrap<edm::Worker::runModule<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)1>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)1>::Context const*)::{lambda()#1}>(edm::Worker::runModule<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)1>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)1>::Context const*)::{lambda()#1}) () from /cvmfs/cms-ib.cern.ch/nweek-02741/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-07-11-2300/lib/slc7_amd64_gcc10/libFWCoreFramework.so
```

The error no longer occurs at the GEM onlineDQM side but at quite a different place. Here is my recipe.

```
# Do not forget voms-proxy-init -voms cms -rfc

export SCRAM_ARCH=slc7_amd64_gcc10
cmsrel CMSSW_12_5_X_2022-07-12-1100
cd CMSSW_12_5_X_2022-07-12-1100/src
cmsenv
git cms-init -q
git cms-merge-topic hyunyong:GEMGeometryBuilder_CMSSW_12_5_X_2022-07-08-2300_PR
git cms-merge-topic francescobrivio:alca-updateDQMrun3
git cms-merge-topic francescobrivio:alca-GEMAlignment
git cms-addpkg DQM/GEM
git cms-addpkg DQM/Integration
scram b -j 6

cmsRun DQM/Integration/python/clients/gem_dqm_sourceclient-live_cfg.py unitTest=True
```

Even I set the CMSSW version (CMSSW_12_5_X_2022-07-12-1100) to be the same as in the log.

Another strange thing happens when I tried the same but with a later version of CMSSW, like CMSSW_12_5_X_2022-07-16-1100, which already includes the two PRs you mentioned. In this case, I obtained the '`NoRecord`' error again. I have no idea what is going on...

**Comment by quark2** [comment]
Hi @mmusich, 

Thanks for your advice. With the merge you gave, the '`NoRecord`' error has been gone. Instead, the `abort signal` runtime error I met just before occurs again... I'm still looking at what happened exactly.

**Comment by tvami** [comment]
hi @hyunyong @jshlee what's the status of this PR?

**Comment by quark2** [comment]
Hi, @tvami, 

Sorry for my late reply. It takes a super long time to catch what causes the problem exactly... Fortunately, I've found it and made a PR to fix it (#38908). Could you test this PR with the new one?

Best regards, 
Byeonghak Ko

**Comment by srimanob** [comment]
test parameters:
- pull_request = https://github.com/cms-sw/cmssw/pull/38908

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by srimanob** [comment]
@hyunyong @jshlee @watson-ij 
I assume this PR is ready to go, right?

And comment https://github.com/cms-sw/cmssw/pull/38666/files#r919032352 will be followed up later.

**Comment by hyunyong** [comment]
> https://github.com/cms-sw/cmssw/pull/38666/files#r919032352

Yes, this PR is ready to go.
Also, yes, we will follow the comment.

**Comment by francescobrivio** [comment]
> > https://github.com/cms-sw/cmssw/pull/38666/files#r919032352
> 
> Yes, this PR is ready to go. Also, yes, we will follow the comment.

Do you plan to also backport https://github.com/cms-sw/cmssw/pull/38908 to 12_4_X?

**Comment by srimanob** [comment]
+Upgrade

**Comment by francescobrivio** [comment]
> Yes. @quark2 please backport your PR.

@hyunyong @quark2  I backported the changes from #38908 in my PR: #38694

**Comment by civanch** [comment]
+1

**Comment by tvami** [comment]
urgent
 * I understand this needs to be used in 12_4_X

**Comment by qliphy** [comment]
please test
to refresh

**Comment by qliphy** [comment]
@hyunyong  Just to confirm that the differences shown in wf 138.5 is expected or not?:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-10-2300+385255/52071/138.5_ExpressCollisions+RunMinimumBias2021+TIER0EXPRUN3+ALCARECOEXPR3+HARVESTDEXPR3/GEM_RecHits.html

**Comment by qliphy** [comment]
> @hyunyong Just to confirm that the differences shown in wf 138.5 is expected or not?: https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-10-2300+385255/52071/138.5_ExpressCollisions+RunMinimumBias2021+TIER0EXPRUN3+ALCARECOEXPR3+HARVESTDEXPR3/GEM_RecHits.html

@jshlee @watson-ij

**Comment by qliphy** [comment]
please test

**Comment by qliphy** [comment]
@hyunyong @jshlee @watson-ij    the differences shown in wf 138.5 is expected or not?:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-14-2300+385255/52125/138.5_ExpressCollisions+RunMinimumBias2021+TIER0EXPRUN3+ALCARECOEXPR3+HARVESTDEXPR3/GEM.html

**Comment by quark2** [comment]
Hi @qliphy, 

Sorry for my late reply. I've reproduced the compared plots, some of which are the follows: 
image
image

The changes are slightly different, which should be due to the alignment. All other 3 plots have similar slight shifts. Note that this PR is for the alignment, and these plots are from real data (the wf 138.5.)

Best regards, 
Byeonghak Ko

**Comment by qliphy** [comment]
please test
to check again

**Comment by qliphy** [comment]
+1

**Pull request merged**